flang/test/Lower/where.f90

```diff
@@ -107,3 +110,28 @@
    ! CHECK-DAG: freemem %[[uvec]]
    ! CHECK: return
 end
+
+! CHECK-LABEL: func @_QPissue855(
+! CHECK-SAME: %[[x:.*]]: !fir.ref<!fir.array<?x?xf32>>
+subroutine issue855(x, n)
+  integer(8) :: n
+  real :: x(n, n)
+  ! CHECK-DAG: fir.array_load %[[x]](%{{.*}}) : (!fir.ref<!fir.array<?x?xf32>>, !fir.shape<2>) -> !fir.array<?x?xf32>
+  ! CHECK-DAG: %[[two:.*]] = constant 2.0{{.*}} : f32
+  ! CHECK: %[[tvec:.*]] = fir.allocmem !fir.array<?x?x!fir.logical
+  ! CHECK-DAG: fir.array_load %[[tvec]]
+  ! CHECK: fir.do_loop
+  ! CHECK: cmpf ogt, %{{.*}}, %[[two]]
+  ! CHECK: fir.array_merge_store %{{.*}}, %{{.*}} to %[[tvec]]
+  ! CHECK: %[[tmask:.*]] = fir.array_load %[[tvec]]
```

Comment:
This looks quite off to me. Why is the array being merged and reloaded in the body of the expression?

Reply:
It's not the body of a single expression, it is the body of the where statement where several things must happen in order.
- the mask is first evaluated in a temp (first fir.array_load, fir.do_loop and fir.array_merge_store)
- and then the result array value is used in the actual assignment inside the where statement (hence the second fir.array_load of the temp).

My understanding of array operations is that it would be incorrect to use the first array_load in the second part of the where statement lowering given it was created when the temp was not initialized with the mask value.

Reply:
Updated to use fir.array_coor.